compiler/src/dotty/tools/dotc/transform/ExplicitSelf.scala

```diff
@@ -26,18 +23,28 @@ class ExplicitSelf extends MiniPhase {
 
   override def phaseName: String = "explicitSelf"
 
+  private def needsCast(tree: RefTree, cls: ClassSymbol)(using Context) =
+    !cls.is(Package) && cls.givenSelfType.exists && !cls.derivesFrom(tree.symbol.owner)
+
+  private def castQualifier(tree: RefTree, cls: ClassSymbol, thiz: Tree)(using Context) =
+    cpy.Select(tree)(thiz.cast(AndType(cls.classInfo.selfType, thiz.tpe)), tree.name)
+
   override def transformIdent(tree: Ident)(using Context): Tree = tree.tpe match {
     case tp: ThisType =>
       report.debuglog(s"owner = ${ctx.owner}, context = ${ctx}")
       This(tp.cls).withSpan(tree.span)
-    case _ => tree
+    case TermRef(thisTp: ThisType, _) =>
+      val cls = thisTp.cls
+      if needsCast(tree, cls) then castQualifier(tree, cls, This(cls))
```

Comment:
```suggestion
      if needsCast(tree, cls) then castQualifier(tree, cls, This(cls).withSpan(tree.span))
```

Reply:
It will get the span anyway through the cpy.Select(tree) call.